explainability-core/src/main/java/org/kie/trustyai/explainability/local/counterfactual/CounterfactualDiversifier.java

```diff
@@ -0,0 +1,236 @@
+
+        return result;
+    };
+    private static final Function<Output, Object> getValueObject = item -> item.getValue().getUnderlyingObject();
+
+    private class FeatureDistance {
+        public double getDistance() {
+            return distance;
+        }
+
+        public Prediction getPrediction() {
+            return prediction;
+        }
+
+        private final double distance;
+        private final Prediction prediction;
+
+        FeatureDistance(Prediction prediction, double distance) {
+            this.prediction = prediction;
+            this.distance = distance;
+        }
+    }
+
+    /**
+     * Instantiate a diverse counterfactual generator.
+     *
+     * @param model The {@link PredictionProvider} required for predicting new counterfactuals
+     * @param original The original reference list of {@link Feature}
+     * @param result The original {@link CounterfactualResult}
+     * @param nSamples The number of samples to create
+     * @param goal The counterfactual's goal as a {@link List<Output>}
+     */
+    public CounterfactualDiversifier(PredictionProvider model,
+            List<Feature> original, CounterfactualResult result, int nSamples, List<Output> goal,
+            BiFunction<List<Feature>, Prediction, Double> distanceCriteria) {
+        this.model = model;
+        this.original = original;
+        this.nFeatures = original.size();
+        this.result = result;
+        this.nSamples = nSamples;
+        this.goal = goal;
+        this.distanceCriteria = distanceCriteria;
+        this.rng = new Random();
+    }
+
+    /**
+     * Generate dataset with diverse counterfactuals.
+     *
+     * @param maximum Maximum number of diverse counterfactuals to generate
+     * @return A {@link Dataset} with the diverse counterfactuals.
+     * @throws ExecutionException
+     * @throws InterruptedException
+     */
+    public Dataset diversify(int maximum) throws ExecutionException, InterruptedException {
+        final Dataset dataset = this.generateAlternatives();
+        final List<Prediction> data = dataset.getData();
+        final List<FeatureDistance> distances = data
+                .stream()
+                .map(d -> new FeatureDistance(d, this.distanceCriteria.apply(original, d)))
+                .sorted(Comparator.comparing(FeatureDistance::getDistance)).collect(Collectors.toList());
+        final List<Prediction> predictions = distances.stream().limit(maximum).map(FeatureDistance::getPrediction).collect(Collectors.toList());
+        return new Dataset(predictions);
+    }
+
+    public static boolean goalMatches(List<Output> output, List<Output> goal) {
+        final List<Object> outputValues = output.stream().map(getValueObject).collect(Collectors.toList());
+        final List<Object> goalValues = goal.stream().map(getValueObject).collect(Collectors.toList());
+
+        return outputValues.equals(goalValues);
+    }
+
+    private double[] generateProbabilities() {
```

Comment:
@RobGeada Assuming CF changed features are sparse + low probability of replacement, we would get a high amount of unchanged candidates. The 50% is bit arbitrary, of course, and considering that we could actually just try deterministic permutations we could just decouple the candidate generation to a pluggable interface and provided a few alternatives (with a default one in case no preference exists) in the future.
IMO the best method will depend on the model type, numbers of features, sparsity, etc.